cpp/src/slice2js/Gen.cpp

```diff
@@ -2293,12 +2293,31 @@ Slice::Gen::TypeScriptRequireVisitor::addImport(const TypePtr& definition, const
 {
     if(!BuiltinPtr::dynamicCast(definition))
     {
-        const string m1 = getModuleMetadata(definition);
+        string m1 = getModuleMetadata(definition);
         const string m2 = getModuleMetadata(toplevel);
 
-        const string p1 = definition->definitionContext()->filename();
+        string p1 = definition->definitionContext()->filename();
         const string p2 = toplevel->definitionContext()->filename();
 
+        ProxyPtr p = ProxyPtr::dynamicCast(definition);
+        ClassDeclPtr c = p ? p->_class() : ClassDeclPtr::dynamicCast(definition);
+
+        if(c)
+        {
+            if(c->definition() == 0)
+            {
+                string definedIn = getDefinedIn(c);
+                if(!definedIn.empty())
+                {
+                    p1 = definedIn;
+                }
+            }
+            else
+            {
+                m1 = getModuleMetadata(ContainedPtr::dynamicCast(c->definition()));
+                p1 = c->definition()->definitionContext()->filename();
+            }
```

Comment:
Do we need this else? When in the else, the class/interface is fully defined, which should be just like before this patch.

Reply:
Without this for a forward declared type that is already defined, we import the file corresponding to the forward declaration and not the one corresponding to the definition. It was bogus before.